src/combinatorics/simple_graph/matching.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2020 Alena Gusakov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alena Gusakov.
+-/
+import data.fintype.basic
+import data.sym2
+import combinatorics.simple_graph.basic
+/-!
+# Matchings
+
+
+## Main definitions
+
+* a `matching` on a simple graph is a subset of its edge set such that
+   no two edges shares an endpoint.
+
+* a `perfect_matching` on a simple graph is a matching in which every
+   vertex belongs to an edge.
+
+TODO:
+  - Lemma stating that the existence of a perfect matching on `G` implies that
+    the cardinality of `V` is even (assuming it's finite)
+  - Hall's Marriage Theorem
+  - Tutte's Theorem
+
+TODO: Tutte and Hall require a definition of subgraphs.
+-/
+open finset
+universe u
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V) (M : set (sym2 V)) (h : M ⊆ edge_set G)
+
+instance : inhabited (set (sym2 V)) :=
+  { default := (complete_graph V).edge_set }
+
+/--
+A matching on `G` is a subset of its edges such that no two edges share a vertex.
+-/
+def matching : Type* :=
+  { M : set (sym2 V) // M ⊆ edge_set G ∧ ∀ x ∈ M, ∀ y ∈ M, ∀ v : V, v ∈ x ∧ v ∈ y → x = y }
+
+instance : inhabited (matching G) := { default :=
+  ⟨(∅ : set (sym2 V)),
+    begin
+      split,
+      exact set.empty_subset G.edge_set,
+      intros x hx y hy v hv,
+      have h2 := set.mem_empty_eq x,
+      by_contra,
+      rw ← h2,
+      exact hx,
+    end ⟩ }
+
+/--
+A perfect matching `M` on graph `G` is a matching such that
+  every vertex is contained in an edge of `M`.
+-/
+def perfect_matching (M : matching G) :
+  Prop := ∀ v : V, ∃ x ∈ M.1, v ∈ x
```

Comment:
Formatting:
```lean
def perfect_matching (M : matching G) : Prop :=
∀ v : V, ∃ x ∈ M.val, v ∈ x
```

Reply:
It might also be useful to split this up:
```lean
def matching_verts (M : G.matching) : set V :=
{v : V | ∃ x ∈ M.val, v ∈ x}

def perfect_matching (M : matching G) : Prop :=
G.matching_verts M = set.univ
```
(I'm imagining it might be useful looking at the lattice of matchings and the corresponding lattice of matched vertices.)

Reply:
Can we also express `matching_verts` as a union?

Reply:
I'm not sure the best way to write it, but here's a start (including a missing `sym2` instance and lemma):
```lean
instance sym2.coe_to_set (α : Type*) : has_coe (sym2 α) (set α) :=
{ coe := λ z, {x | x ∈ z} }

@[simp] lemma sym2.mem_coe {α : Type*} {x : α} {z : sym2 α} (h : x ∈ (z : set α)) : x ∈ z :=
h

def matching_verts (M : G.matching) : set V :=
⋃₀(set.image coe M.val)
```
(Maybe it should be `has_lift` instead of `has_coe`? I'm going by the idea that things with `has_mem` "are" sets.)

Reply:
@awainverse are you saying union as in the union over the edge set in the matching?

Reply:
Thinking of the matching as a set of sets or finsets of size 2, then `matching_verts` (which should maybe be called the `support` or something) can be viewed as the union of all those set/finsets. I'm not sure if it's actually worth developing that coe in this PR to do that though, I kind of forgot that they weren't already finsets.

Reply:
Right, that makes sense. I think I'll hold onto that thought for now and maybe show that the two are definitionally equivalent at some point or something